Action Goal Preemption Bug Fix

vittoriouss opened the pull request:
This PR Does the following changes to mongodb_play.py: 

1. now queue_from_db checks if 'goal_id' field exists in the document

2. If the message type is an actionGoal (contains 'goal_id') before publishing the message the goal_id time stamp will be changed to current wall clock time. This allows for the replayed goal to
preempt any previous goals. For more details on action lib msg type goal_id look [here](http://docs.ros.org/fuerte/api/actionlib_msgs/html/msg/GoalID.html)

vittoriouss committed:
This commit does the following changes:

1. now queue_from_db checks for msg_type (only once)

2. If the message type is an actionGoal (contains substring 'Goal')
before publishing the message the goal_id time stamp will be changed
to current wall clock time. This allows for the replayed goal to
preempt any previous goals.

vittoriouss commented on mongodb_store/scripts/mongodb_play.py:
```diff
@@ -156,9 +156,17 @@ def run(self, running):
+        try:
+            action_goal_match = True if "Goal" in self.msg_type else False
+            msg_type_not_init = False
+        except AttributeError as e:
```
sometimes self.msg_type is not initialized before this runs (async issue?). Not the best way to catch this issue, but it works fairly quickly. I would rather not do a substring match every single time we publish a topic

umhan35 commented on mongodb_store/scripts/mongodb_play.py:
```diff
@@ -120,7 +120,7 @@ def queue_from_db(self, running):
         # load message class for this collection, they should all be the same
         msg_cls = mg_util.load_class(documents[0]["_meta"]["stored_class"])
-
+        self.msg_type = documents[0]["_meta"]["stored_type"]
```
maybe add a comment saying that msg_type stay the same throughout the TopicPlayer

vittoriouss replied:
Talked about --- instead of message type check for goal_id attribute

umhan35 commented on mongodb_store/scripts/mongodb_play.py:
```diff
@@ -156,9 +156,17 @@ def run(self, running):
 
         timeout = 1
+        try:
+            action_goal_match = True if "Goal" in self.msg_type else False
```
do a exact match rather than "Goal" to avoid other message types had Goal in their names

umhan35 commented on mongodb_store/scripts/mongodb_play.py:
```diff
@@ -156,9 +156,17 @@ def run(self, running):
         timeout = 1
+        try:
+            action_goal_match = True if "Goal" in self.msg_type else False
+            msg_type_not_init = False
```
comment on the purpose of this variable

umhan35 commented on mongodb_store/scripts/mongodb_play.py:
```diff
@@ -156,9 +156,17 @@ def run(self, running):
         while running.value:
             try:
+                if msg_type_not_init:
+                    action_goal_match = True if "Goal" in self.msg_type else False
```
same here

umhan35 commented on mongodb_store/scripts/mongodb_play.py:
```diff
@@ -171,6 +179,12 @@ def run(self, running):
                 else:
                     delay = publish_time - now
                     rospy.sleep(delay)
+                if action_goal_match:
```
an empty line before the if block

umhan35 commented on mongodb_store/scripts/mongodb_play.py:
```diff
@@ -156,9 +156,17 @@ def run(self, running):
         self.event.wait()
 
         timeout = 1
+        try:
```
leave a comment explaining why we need this and how this problem is solved

umhan35 commented on mongodb_store/scripts/mongodb_play.py:
```diff
@@ -171,6 +179,12 @@ def run(self, running):
+                if action_goal_match:
+                    float_secs = time.time()
+                    secs = int(float_secs)
+                    nsecs = int((float_secs - secs) * 1000000000)
```
explain why these values are set, the `1000000000`

umhan35 replied:
did 0 work?

vittoriouss replied:
This was copied from rospy on how they get the wall clock time.

vittoriouss replied:
I added a comment saying where I got these lines of code frm

vittoriouss committed:
PR comments

1. Instead of checking if the message type contains a the substring
goal the queue_from_db checks if the goal_id exists in the document

2. if the goal id does exist then the message's goal id time stamp
will change to reflect the current wall clock time

vittoriouss committed:
Added a more efficient way of checking for the existance of 'goal-id'
Time complexity should be currently O(1) (before it was O(n))

vittoriouss committed:
added empty line before if block

umhan35 commented on mongodb_store/scripts/mongodb_play.py:
```diff
@@ -172,6 +173,13 @@ def run(self, running):
+                    float_secs = time.time()
+                    secs = int(float_secs)
+                    msg.goal_id.stamp.secs = secs
+                    msg.goal_id.stamp.nsecs = int((float_secs - secs) * 1000000000)
```
I think you can use `msg.goal_id.stamp = ` `rospy.Time.now()` or `rospy.get_rostime()`?

doc: http://wiki.ros.org/rospy/Overview/Time#Getting_the_current_time

```
now = rospy.get_rostime()
rospy.loginfo("Current time %i %i", now.secs, now.nsecs)
```

umhan35 replied:
I suggested `msg.goal_id.stamp = 0`, but not sure if it works...

vittoriouss replied:
That returns the current sim time. This is what the definition of what Time.now() says: 
```
Create new L{Time} instance representing current time. This 
140          can either be wall-clock time or a simulated clock. It is 
141          strongly recommended that you use the now() factory to create 
142          current time representations instead of reading wall-clock 
143          time and create Time instances from it. 
144           
145          @return: L{Time} instance for current time 
146          @rtype: L{Time} 
```
Time.now() will return the simulated clock time, which is different than the wall clock time. If we set the goal_id stamp to the simulated clock time it'll be the same value as before. The wallclock time will allow for no issues.
